File: src/0218-The-Skyline-Problem/0218.go
type IntHeap [][]int

func (h IntHeap) Len() int {
    return len(h)
}

func (h IntHeap) Less(i, j int) bool {
    if h[i][0] == h[j][0] {
        return h[i][1] < h[j][1]
    }
    return h[i][0] < h[j][0]
}

func (h IntHeap) Swap(i, j int) {
    h[i], h[j] = h[j], h[i]
}

func (h *IntHeap) Pop() interface{} {  
    x := (*h)[(*h).Len()-1]
	*h = (*h)[:(*h).Len()-1]
    return x
}

func (h *IntHeap) Push(x interface{}) {  
    *h = append(*h, x.([]int))
}

func getSkyline(buildings [][]int) [][]int {
    const INT_MAX = int(^uint(0) >> 1)
    points := make([][]int, 0)
    rSet := make(map[int]int)
    for _, v := range buildings {
        points = append(points, []int{v[0], -v[2], v[1]})
        if _, ok := rSet[v[1]]; !ok {
            points = append(points, []int{v[1], 0, 0})
            rSet[v[1]] = 1
        }
    }
    sort.Slice(points, func(a, b int) bool {
        if points[a][0] == points[b][0] {
            return points[a][1] < points[b][1]
        }
        return points[a][0] < points[b][0]
    })
    h := new(IntHeap)
    heap.Push(h, []int{0, INT_MAX})
    res := [][]int{{0, 0}}
    for _, v := range points {
        for (*h)[0][1] <= v[0] {
            heap.Pop(h)
        }
        if v[1] != 0 {
            heap.Push(h, []int{v[1], v[2]})
        }
        if res[len(res) - 1][1] != -(*h)[0][0] {
            res = append(res, []int{v[0], -(*h)[0][0]})
        }
    }
    return res[1:len(res)]
}

File: src/0218-The-Skyline-Problem/0218.js
const t = 0;
const parent = i => ((i + 1) >>> 1) - 1;
const left = i => (i << 1) + 1;
const right = i => (i + 1) << 1;

class PriorityQueue {
    constructor(comparator = (a, b) => a > b) {
        this._heap = [];
        this._comparator = comparator;
    }

    size() {
        return this._heap.length;
    }

    isEmpty() {
        return this.size() == 0;
    }

    peek() {
        return this._heap[t];
    }

    push(...values) {
        values.forEach(value => {
            this._heap.push(value);
            this._siftUp();
        });
        return this.size();
    }

    pop() {
        const poppedValue = this.peek();
        const bottom = this.size() - 1;
        if (bottom > t) {
            this._swap(t, bottom);
        }
        this._heap.pop();
        this._siftDown();
        return poppedValue;
    }

    replace(value) {
        const replacedValue = this.peek();
        this._heap[t] = value;
        this._siftDown();
        return replacedValue;
    }

    _greater(i, j) {
        return this._comparator(this._heap[i], this._heap[j]);
    }

    _swap(i, j) {
        [this._heap[i], this._heap[j]] = [this._heap[j], this._heap[i]];
    }

    _siftUp() {
        let node = this.size() - 1;
        while (node > t && this._greater(node, parent(node))) {
            this._swap(node, parent(node));
            node = parent(node);
        }
    }

    _siftDown() {
        let node = t;
        while (
            (left(node) < this.size() && this._greater(left(node), node)) ||
            (right(node) < this.size() && this._greater(right(node), node))
        ) {
            let maxChild = (right(node) < this.size() && this._greater(right(node), left(node))) ? right(node) : left(node);
            this._swap(node, maxChild);
            node = maxChild;
        }
    }
}
/**
 * @param {number[][]} buildings
 * @return {number[][]}
 */
var getSkyline = function(buildings) {
    let points = [], rSet = new Set();
    for (let it of buildings) {
        points.push([it[0], -it[2], it[1]]);
        if (!rSet.has(it[1])) {
            points.push([it[1], 0, 0])
            rSet.add(it[1]);
        }
    }
    let cmp = (a, b) => {
        if (a[0] == b[0]) return a[1] - b[1];
        return a[0] - b[0];
    }
    points.sort(cmp);
    const q = new PriorityQueue(cmp);
    q.push([0, Number.MAX_VALUE])
    let res = [[0, 0]];
    for (let it of points) {
        while (q.peek()[1] <= it[0]) q.pop();
        if (it[1] != 0) q.push([it[1], it[2]]);
        if (res[res.length-1][1] != -q.peek()[0])
            res.push([it[0], -q.peek()[0]]);
    }
    res.shift();
    return res;
};